src/_macosx.m

```diff
@@ -178,6 +178,101 @@ static int wait_for_stdin(void)
     return 1;
 }
 
+static PyObject* set_icon(PyObject* ignored, PyObject* path)
+{
+    if (!path)
+    {
+        PyErr_SetString(PyExc_RuntimeError, "No path passed");
+        return NULL;
+    }
+
+    const char* path_string;
+    PyObject* unicode_path = PyObject_Str(path);
```

Comment:
personally I prefer a style like https://github.com/matplotlib/matplotlib/blob/master/src/_tkagg.cpp#L306 (more code, less blank lines...) though it's just my opinion.

Reply:
Also, probably this should use https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_DecodeFSDefault instead of PyObject_Str?

Reply:
I can’t use that one- the input arguments are different, `PyObject_Str` takes a `PyObject`, but `PyUnicode_DecodeFSDefault` takes a `const char*`.

Reply:
But you still need to perform proper decoding...
I guess with any of these, don't try to retrieve the icon path in C code at all; just expose a function that takes a bytes as argument (the decoded path) and sets the icon from it, and call that function from the python-level backend_macos.py.

Reply:
I think that's worse, the bytes could be anything and might not decode as proper UTF8. The NSImage API doesn't document what happens if I pass in invalid UTF8...

Working backwards:

- I'd like to keep the Python decoding next to the NSImage API. Python `PyUnicode_AsUTF8 ` API says it'll give me a UTF-8 and NSImage needs a UTF-8.
- The Python `PyUnicode_AsUTF8` takes PyObject*, so I'd like to give it a `str`. It'll fail on anything else, including things like `PosixPath`.

So I think I have two choices:

- C function takes any objects and converts them to `str` somehow
- C function only works if you give it a `str`, fails on anything else

I wish I could do the first without converting everything or doing type-checking. E.g., `str(None)` becomes "None".

How would you do this?

Reply:
Then make your function take a str as argument and decode it using the `s` code per https://docs.python.org/3/c-api/arg.html#strings-and-buffers (which will decode with utf8 into a const char*).

Reply:
Ooooo this note on `s` is helpful:

> Note This format does not accept bytes-like objects. If you want to accept filesystem paths and convert them to C character strings, it is preferable to use the O& format with PyUnicode_FSConverter() as converter.

That's exactly what I want to do. I'll fix it using Python's recommendation, yeah?

Reply:
yes

Reply:
See latest push.